### PR TITLE
ci: Add "macOS 12 native arm64" task

### PR DESCRIPTION
**Pull request by hebasto**
The release binaries for Apple Silicon macOS (`arm64-apple-darwin`) have been available since v23.0.

It seems reasonable to add a task for such a platform to CI. Also it would prevent regressions like bitcoin/bitcoin#24958.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Would this actually prevent #24958? Wouldn't it silently "fail" green?

**Comment by hebasto** [comment]
> Would this actually prevent #24958? Wouldn't it silently "fail" green?

https://github.com/bitcoin/bitcoin/blob/f2e07bdbac58c2cd38cd5f668847e95cde3eeefd/ci/test/00_setup_env_mac_native_arm64.sh#L11

It will fail due to https://github.com/bitcoin/bitcoin/blob/f7a1e676d536115aba649672ba1e68755a1c90dc/configure.ac#L1751-L1757

**Comment by Zero-1729** [comment]
Concept ACK

**Comment by MarcoFalke** [comment]
Not sure if it makes sense to run this check on every commit. There is some probability of breaking that is acceptable and unavoidable. I think we should focus on checks that are likely to catch errors often enough to be a net-positive considering the maintenance overhead and CPU overhead of the added CI task.

So if https://github.com/bitcoin/bitcoin/pull/24958 is the only known bug this prevents, I tend toward NACK, even though the code looks good and correct.

I expect that there are ppl running M1 locally, but if not, this could also be implemented as a nightly task to run outside this repo.

File: ci/test/04_install.sh [hunk 1]
```diff
@@ -10,10 +10,10 @@ if [[ $QEMU_USER_CMD == qemu-s390* ]]; then
   export LC_ALL=C
 fi
 
-if [ "$CI_OS_NAME" == "macos" ]; then
+if [ "$CI_OS_NAME" == "macos" ] && [ -n "$PIP_PACKAGES" ]; then
   sudo -H pip3 install --upgrade pip
   # shellcheck disable=SC2086
-  IN_GETOPT_BIN="/usr/local/opt/gnu-getopt/bin/getopt" ${CI_RETRY_EXE} pip3 install --user $PIP_PACKAGES
+  IN_GETOPT_BIN="$(brew --prefix gnu-getopt)/bin/getopt" ${CI_RETRY_EXE} pip3 install --user $PIP_PACKAGES
```
Comment (promag) [inline ci/test/04_install.sh:16]: Nice one 🤓

File: .cirrus.yml [hunk 2]
```diff
@@ -306,18 +314,28 @@ task:
...
     image: monterey-xcode-13.3  # https://cirrus-ci.org/guide/macOS
+  << : *MACOS_NATIVE_TASK_TEMPLATE
+  env:
+    << : *CIRRUS_EPHEMERAL_WORKER_TEMPLATE_ENV
+    CI_USE_APT_INSTALL: "no"
+    PACKAGE_MANAGER_INSTALL: "echo"  # Nothing to do
+    FILE_ENV: "./ci/test/00_setup_env_mac_native_x86_64.sh"
+
+task:
+  name: 'macOS 12 native Apple Silicon [gui, system sqlite] [no depends]'
```
Comment (promag) [inline .cirrus.yml:329]: nit, s/Apple Silicon/arm64

**Review by promag — COMMENTED** [review]
Concept ACK

**Comment by hebasto** [comment]
> CPU overhead of the added CI task

I don't think the M1 CPU time is shared with any other task.

**Comment by MarcoFalke** [comment]
According to https://cirrus-ci.org/faq/#are-there-any-limits  there is a limit of "12.0 CPUs macOS VM (1 VM)", so it will be shared with the other macos instance?

**Comment by hebasto** [comment]
> According to https://cirrus-ci.org/faq/#are-there-any-limits there is a limit of "12.0 CPUs macOS VM (1 VM)", so it will be shared with the other macos instance?

It looks like docs are a bit inconsistent after introducing Apple Silicon support.

They are correct about [`x86_64`](https://cirrus-ci.com/task/5473695194939392):
image

But M1 limits [differ](https://cirrus-ci.com/task/6599595101782016):
image

**Comment by MarcoFalke** [comment]
Ok, looks like it won't affect the limits for now. Though, I am still wondering how many bugs this will catch that are not found by other tasks.

**Comment by hebasto** [comment]
Updated f2e07bdbac58c2cd38cd5f668847e95cde3eeefd -> 84cfbc7329e92191e391ba5bb1991108229bbb51 ([pr25160.02](https://github.com/hebasto/bitcoin/commits/pr25160.02) -> [pr25160.03](https://github.com/hebasto/bitcoin/commits/pr25160.03), [diff](https://github.com/hebasto/bitcoin/compare/pr25160.02..pr25160.03)):

- addressed @promag's [comment](https://github.com/bitcoin/bitcoin/pull/25160#discussion_r876177791)

> I am still wondering how many bugs this will catch that are not found by other tasks.

I hope every bug on the macOS `arm64` platform, which has a steadily growing number of users, will be caught before release.

File: .cirrus.yml (hunk 2 above)
Comment (hebasto) [reply .cirrus.yml]: [Updated](https://github.com/bitcoin/bitcoin/pull/25160#issuecomment-1135487448).

**Review by Zero-1729 — COMMENTED** [review]
crACK 84cfbc7329e92191e391ba5bb1991108229bbb51

**Comment by MarcoFalke** [comment]
> I hope every bug on the macOS arm64 platform, which has a steadily growing number of users, will be caught before release.

It only caught one bug for now, which doesn't seem enough to justify the maintenance burden? Note that in this repo, *bsd, s390x, various linux distros, valgrind, ... aren't tested either.

**Comment by MarcoFalke** [comment]
I guess the 1st+3rd commit could still make sense. Then I could add it to my btc_nightly CI run, if you want.

**Comment by MarcoFalke** [comment]
if amd64 is unsupported, we should probably remove it and just switch to arm64?

**Pull request closed without merging**